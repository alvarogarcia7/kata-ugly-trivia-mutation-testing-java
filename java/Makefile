.PHONY: mutate
mutate:
	./mvnw clean package org.pitest:pitest-maven:mutationCoverage

