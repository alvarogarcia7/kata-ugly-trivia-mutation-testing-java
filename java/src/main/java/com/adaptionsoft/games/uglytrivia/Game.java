package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;

public class Game {

	private Players playerList = new Players(this);
    ArrayList players = new ArrayList();

    LinkedList popQuestions = new LinkedList();
    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();
    
    int currentPlayer = 0;
    boolean isGettingOutOfPenaltyBox;
	private String verbosity;

	public  Game(){
    	for (int i = 0; i < 50; i++) {
			popQuestions.addLast("Pop Question " + i);
			scienceQuestions.addLast(("Science Question " + i));
			sportsQuestions.addLast(("Sports Question " + i));
		    rockQuestions.addLast("Rock Question " + i);
		}
	}

	public void add(String playerName) {
		
		
	    players.add(playerName);

	    log(playerName + " was added");
	    log("They are player number " + howManyPlayers());
	}

	public int howManyPlayers() {
		return players.size();
	}

	public void roll(int roll) {
		log(players.get(currentPlayer) + " is the current player");
		log("They have rolled a " + roll);

		if (playerList.isCurrentPlayerInPenaltyBox()) {
			if (roll % 2 != 0) {
				isGettingOutOfPenaltyBox = true;

				log(players.get(currentPlayer) + " is getting out of the penalty box");
				movePlayerAndAskQuestion(roll);
			} else {
				log(players.get(currentPlayer) + " is not getting out of the penalty box");
				isGettingOutOfPenaltyBox = false;
			}

		} else {

			movePlayerAndAskQuestion(roll);
		}

	}

	private void movePlayerAndAskQuestion(int roll) {
		playerList.currentPlayerRoll(roll);

		log(players.get(currentPlayer)
			+ "'s new location is "
			+ playerList.currentPlayerPlace());
		log("The category is " + currentCategory());
		askQuestion();
	}

	private void askQuestion() {
		if (currentCategory() == "Pop")
			log(popQuestions.removeFirst());
		if (currentCategory() == "Science")
			log(scienceQuestions.removeFirst());
		if (currentCategory() == "Sports")
			log(sportsQuestions.removeFirst());
		if (currentCategory() == "Rock")
			log(rockQuestions.removeFirst());
	}


	private String currentCategory() {
		if (playerList.currentPlayerPlace() == 0) return "Pop";
		if (playerList.currentPlayerPlace() == 4) return "Pop";
		if (playerList.currentPlayerPlace() == 8) return "Pop";
		if (playerList.currentPlayerPlace() == 1) return "Science";
		if (playerList.currentPlayerPlace() == 5) return "Science";
		if (playerList.currentPlayerPlace() == 9) return "Science";
		if (playerList.currentPlayerPlace() == 2) return "Sports";
		if (playerList.currentPlayerPlace() == 6) return "Sports";
		if (playerList.currentPlayerPlace() == 10) return "Sports";
		return "Rock";
	}

	public boolean wasCorrectlyAnswered() {
		if (playerList.isCurrentPlayerInPenaltyBox()) {
			if (isGettingOutOfPenaltyBox) {
				log("Answer was correct!!!!");
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				playerList.currentPlayerWinsACoin();
				System.out.println(players.get(currentPlayer)
					+ " now has "
					+ playerList.currentPlayerCoins()
					+ " Gold Coins.");

				boolean winner = didPlayerWin();

				return winner;
			} else {
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				return true;
			}


		} else {

			log("Answer was corrent!!!!");
			playerList.currentPlayerWinsACoin();
			System.out.println(players.get(currentPlayer)
				+ " now has "
				+ playerList.currentPlayerCoins()
				+ " Gold Coins.");

			boolean winner = didPlayerWin();
			currentPlayer++;
			if (currentPlayer == players.size()) currentPlayer = 0;

			return winner;
		}
	}

	public boolean wrongAnswer() {
		log("Question was incorrectly answered");
		log(players.get(currentPlayer) + " was sent to the penalty box");
		playerList.currentPlayerToPenaltyBox();

		currentPlayer++;
		if (currentPlayer == players.size()) currentPlayer = 0;
		return true;
	}


	private boolean didPlayerWin() {
		return !(playerList.currentPlayerCoins() == 6);
	}

	public void setVerbosity(String value) {
		this.verbosity = value;
	}

	private void log(Object line) {
		if ("debug".equals(verbosity)) {
			System.out.println(line);
		}
	}


}
