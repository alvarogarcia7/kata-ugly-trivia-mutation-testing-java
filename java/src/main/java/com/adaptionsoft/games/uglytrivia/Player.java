package com.adaptionsoft.games.uglytrivia;

public class Player {
	public final String name;

	private int coins;
	private int place;

	private boolean inPenaltyBox;
	public Player(String name) {
		this.name = name;
		this.coins = 0;
		this.place = 0;
	}

	public int winCoin() {
		return this.coins++;
	}

	public void roll(int roll) {
		place = place + roll;
		if (place > 11) place = place - 12;
	}

	public void toPenaltyBox() {
		inPenaltyBox = true;
	}

	public int getPlace() {
		return place;
	}

	public int getCoins() {
		return coins;
	}

	public boolean isInPenaltyBox() {
		return inPenaltyBox;
	}
}
