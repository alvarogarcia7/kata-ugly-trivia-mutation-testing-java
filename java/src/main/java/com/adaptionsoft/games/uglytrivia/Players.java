package com.adaptionsoft.games.uglytrivia;

public class Players {
	private Game game;
	private int[] purses = new int[6];
	private int[] places = new int[6];
	private boolean[] inPenaltyBox = new boolean[6];


	public Players(Game game) {
		this.game = game;
	}

	int currentPlayerCoins() {
		return purses[game.getCurrentPlayerAsInt()];
	}

	int currentPlayerWinsACoin() {
		return purses[game.getCurrentPlayerAsInt()]++;
	}

	int currentPlayerPlace() {
		return places[game.getCurrentPlayerAsInt()];
	}

	void currentPlayerRoll(int roll) {
		places[game.getCurrentPlayerAsInt()] = currentPlayerPlace() + roll;
		if (currentPlayerPlace() > 11) places[game.getCurrentPlayerAsInt()] = currentPlayerPlace() - 12;
	}

	void currentPlayerToPenaltyBox() {
		inPenaltyBox[game.getCurrentPlayerAsInt()] = true;
	}

	boolean isCurrentPlayerInPenaltyBox() {
		return inPenaltyBox[game.getCurrentPlayerAsInt()];
	}
}
