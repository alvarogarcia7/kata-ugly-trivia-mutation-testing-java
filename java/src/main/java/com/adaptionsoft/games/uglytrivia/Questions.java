package com.adaptionsoft.games.uglytrivia;

import java.util.LinkedList;

public class Questions {
	private LinkedList<String> popQuestions = new LinkedList<>();
	private LinkedList<String> scienceQuestions = new LinkedList<>();
	private LinkedList<String> sportsQuestions = new LinkedList<>();
	private LinkedList<String> rockQuestions = new LinkedList<>();

	public Questions() {
		for (int i = 0; i < 50; i++) {
			popQuestions.addLast("Pop Question " + i);
			scienceQuestions.addLast(("Science Question " + i));
			sportsQuestions.addLast(("Sports Question " + i));
			rockQuestions.addLast("Rock Question " + i);
		}
	}

	public String getPopQuestion() {
		return popQuestions.pop();
	}

	public String getScienceQuestion() {
		return scienceQuestions.pop();
	}

	public String getSportsQuestion() {
		return sportsQuestions.pop();
	}

	public String getRockQuestion() {
		return rockQuestions.pop();
	}

}
