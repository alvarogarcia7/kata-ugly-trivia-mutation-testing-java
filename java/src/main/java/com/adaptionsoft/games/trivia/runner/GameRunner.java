
package com.adaptionsoft.games.trivia.runner;
import java.util.Random;

import com.adaptionsoft.games.uglytrivia.Game;
import com.adaptionsoft.games.uglytrivia.Players;
import com.adaptionsoft.games.uglytrivia.Questions;


public class GameRunner {

	private static boolean notAWinner;

	public static void main(String[] args) {
		Random rand = new Random();
		playGame(rand, "debug");
	}

	public static void playGame(Random rand, String verbosity) {
		Game aGame = new Game(new Questions(), new Players("Chet", "Pat", "Sue"), verbosity);

		do {

			aGame.roll(rand.nextInt(5) + 1);

			if (rand.nextInt(9) == 7) {
				notAWinner = aGame.wrongAnswer();
			} else {
				notAWinner = aGame.wasCorrectlyAnswered();
			}



		} while (notAWinner);
	}
}
