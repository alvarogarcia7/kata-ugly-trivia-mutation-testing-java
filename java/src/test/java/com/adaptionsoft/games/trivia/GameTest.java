package com.adaptionsoft.games.trivia;

import com.adaptionsoft.games.trivia.runner.GameRunner;
import com.adaptionsoft.games.uglytrivia.Game;
import com.adaptionsoft.games.uglytrivia.Players;
import com.adaptionsoft.games.uglytrivia.Questions;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.io.*;
import java.util.Random;
import java.util.stream.IntStream;

import static org.junit.Assert.fail;

public class GameTest {

	@Test
	public void itsLockedDown() throws Exception {

		Random randomizer = new Random(123455);
		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 15).forEach(i -> GameRunner.playGame(randomizer, "debug"));

		Approvals.verify(resultStream.toString());

	}

	@Test
	public void itsLockedDown_longerTest() throws Exception {

		Random randomizer = new Random(123455);
		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 1000).forEach(i -> GameRunner.playGame(randomizer, "debug"));

		Approvals.verify(resultStream.toString());

	}

	@Test
	public void itsLockedDown_broaderTest() throws Exception {

		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 1000).forEach(i -> GameRunner.playGame(new Random(i), "debug"));

		Approvals.verify(resultStream.toString());

	}


	@Test
	public void itsLockedDown_silent() throws Exception {

		Random randomizer = new Random(123455);
		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 15).forEach(i -> GameRunner.playGame(randomizer, ""));

		Approvals.verify(resultStream.toString());

	}

	@Test
	public void itsLockedDown_longerTest_silent() throws Exception {

		Random randomizer = new Random(123455);
		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 1000).forEach(i -> GameRunner.playGame(randomizer, ""));

		Approvals.verify(resultStream.toString());

	}

	@Test
	public void itsLockedDown_broaderTest_silent() throws Exception {

		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		IntStream.range(1, 1000).forEach(i -> GameRunner.playGame(new Random(i), ""));

		Approvals.verify(resultStream.toString());

	}

	@Test
	public void exhaust_all_questions() throws Exception {

		ByteArrayOutputStream resultStream = new ByteArrayOutputStream();
		System.setOut(new PrintStream(resultStream));

		Game aGame = new Game(new Questions(), new Players("Chet", "Pat", "Sue"), "");

		int i = 50;
		int steps = 0;
		do {
			aGame.roll(0);
			aGame.wasCorrectlyAnswered();
			steps++;
			System.out.println(String.format("steps = %d is correct", steps));
			i--;
		} while (i > 0);

		Exception exception = null;
		try {
			aGame.roll(0);
		} catch (Exception e) {
			exception = e;
		} finally {
			if (null == exception) {
				fail("Should have thrown an exception here ");
			}
		}

		Approvals.verify(resultStream.toString());

	}

}
