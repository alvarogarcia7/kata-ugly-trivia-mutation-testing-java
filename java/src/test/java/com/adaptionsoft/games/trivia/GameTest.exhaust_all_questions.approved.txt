Chet now has 1 Gold Coins.
steps = 1 is correct
Pat now has 1 Gold Coins.
steps = 2 is correct
Sue now has 1 Gold Coins.
steps = 3 is correct
Chet now has 2 Gold Coins.
steps = 4 is correct
Pat now has 2 Gold Coins.
steps = 5 is correct
Sue now has 2 Gold Coins.
steps = 6 is correct
Chet now has 3 Gold Coins.
steps = 7 is correct
Pat now has 3 Gold Coins.
steps = 8 is correct
Sue now has 3 Gold Coins.
steps = 9 is correct
Chet now has 4 Gold Coins.
steps = 10 is correct
Pat now has 4 Gold Coins.
steps = 11 is correct
Sue now has 4 Gold Coins.
steps = 12 is correct
Chet now has 5 Gold Coins.
steps = 13 is correct
Pat now has 5 Gold Coins.
steps = 14 is correct
Sue now has 5 Gold Coins.
steps = 15 is correct
Chet now has 6 Gold Coins.
steps = 16 is correct
Pat now has 6 Gold Coins.
steps = 17 is correct
Sue now has 6 Gold Coins.
steps = 18 is correct
Chet now has 7 Gold Coins.
steps = 19 is correct
Pat now has 7 Gold Coins.
steps = 20 is correct
Sue now has 7 Gold Coins.
steps = 21 is correct
Chet now has 8 Gold Coins.
steps = 22 is correct
Pat now has 8 Gold Coins.
steps = 23 is correct
Sue now has 8 Gold Coins.
steps = 24 is correct
Chet now has 9 Gold Coins.
steps = 25 is correct
Pat now has 9 Gold Coins.
steps = 26 is correct
Sue now has 9 Gold Coins.
steps = 27 is correct
Chet now has 10 Gold Coins.
steps = 28 is correct
Pat now has 10 Gold Coins.
steps = 29 is correct
Sue now has 10 Gold Coins.
steps = 30 is correct
Chet now has 11 Gold Coins.
steps = 31 is correct
Pat now has 11 Gold Coins.
steps = 32 is correct
Sue now has 11 Gold Coins.
steps = 33 is correct
Chet now has 12 Gold Coins.
steps = 34 is correct
Pat now has 12 Gold Coins.
steps = 35 is correct
Sue now has 12 Gold Coins.
steps = 36 is correct
Chet now has 13 Gold Coins.
steps = 37 is correct
Pat now has 13 Gold Coins.
steps = 38 is correct
Sue now has 13 Gold Coins.
steps = 39 is correct
Chet now has 14 Gold Coins.
steps = 40 is correct
Pat now has 14 Gold Coins.
steps = 41 is correct
Sue now has 14 Gold Coins.
steps = 42 is correct
Chet now has 15 Gold Coins.
steps = 43 is correct
Pat now has 15 Gold Coins.
steps = 44 is correct
Sue now has 15 Gold Coins.
steps = 45 is correct
Chet now has 16 Gold Coins.
steps = 46 is correct
Pat now has 16 Gold Coins.
steps = 47 is correct
Sue now has 16 Gold Coins.
steps = 48 is correct
Chet now has 17 Gold Coins.
steps = 49 is correct
Pat now has 17 Gold Coins.
steps = 50 is correct
